### ISSUE TITLE
simple fuzzer is weak against len+data fields

### ISSUE THREAD

**Issue by AlekseyCherepanov** [openwall/john#5147]
#5086 fixed overflow in PEM format. Built-in simple fuzzer could not find this problem.

```
$ python -c 'l = 8000; print "$PEM$1$1$0201000282010100$2048$308204bd02010030$" + str(l) + "$" + "a" * (l * 2)' > t.pw

$ ../run/john t.pw
[...]
==24239==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x62100002693f at pc 0x560d44b28c24 bp 0x7ffd702644c0 sp 0x7ffd702644b8
WRITE of size 1 at 0x62100002693f thread T0
    #0 0x560d44b28c23 in pem_get_salt /home/user/john/src/pem_common_plug.c:132
    #1 0x55eb63c45546 in ldr_load_pw_line /home/user/john/src/loader.c:1050
[...]

$ python -c 'l = 4097; print "$PEM$1$1$0201000282010100$2048$308204bd02010030$" + str(l) + "$" + "a" * (l * 2)' > t.pw

$ ../run/john t.pw
[...]
==24734==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffe1d33c6c0 at pc 0x55ed2c8de033 bp 0x7ffe1d33b1e0 sp 0x7ffe1d33b1d8
READ of size 1 at 0x7ffe1d33c6c0 thread T0
    #0 0x55ed2c8de032 in check_pkcs_pad /home/user/john/src/jumbo.c:432
    #1 0x55ed2cc43f48 in pem_decrypt /home/user/john/src/pem_common_plug.c:203
    #2 0x55ed2cc45981 in crypt_all /home/user/john/src/pem_fmt_plug.c:124
[...]
```

Big overflow gets caught by ASan early. But small overflow hits later because overflow happens inside structure.

`pem_common.h`:
```c
struct custom_salt {
	int salt_length;
	unsigned char salt[SALTLEN];
	int iv_length;
	unsigned char iv[16];
	int iterations;
	int ciphertext_length;
	unsigned char ciphertext[CTLEN];
	int cid;  // cipher id
	int key_length;
};
```

`pem_common_plug.c`:
```c
// $PEM$type$cipher$$salt$iterations$iv$blob_length$blob  // type, and cipher should be enough for all possible combinations
```

The overflow happens in the `blob` field. But it requires respective `blog_length` to happen.

Overflow happens decoding hex in `pem_get_salt`:
```c
	cur_salt->ciphertext_length = atoi(p);
	p = strtokm(NULL, "$");
	for (i = 0; i < cur_salt->ciphertext_length; i++)
		cur_salt->ciphertext[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16
			+ atoi16[ARCH_INDEX(p[i * 2 + 1])];
```

I tried to set `ciphertext_length` to `strlen(p) / 2` right before the cycle. With such change, simple fuzzer triggers overflow.

There are a lot of formats with redundant length fields. We might want to remove these fields. Saving compatibility, we might just ignore them completely instead of rejection when they are not in sync with respective data fields.

I guess AFL / libFuzzer can trigger the overflow.

I guess wordlist of simple fuzzer could be improved to trigger the overflow. I'll investigate it.

**Comment by AlekseyCherepanov** [comment]
> I guess wordlist of simple fuzzer could be improved to trigger the overflow. I'll investigate it.

Each string from a dictionary for fuzzing is inserted at different positions (at each position for short ciphertext, only before and after delimiters in very long ciphertexts). So the idea is to insert length, delimiter, hex blob of respective length, delimiter. It works but there are problems.

For `pem`, it turned out that the second delimiter should be `*` instead of its regular `$`:
```c
	if ((p = strtokm(NULL, "$")) == NULL)   // ciphertext length
		goto err;
	if (!isdec(p))
		goto err;
	len = atoi(p);
	if ((p = strtokm(NULL, "*")) == NULL)   // ciphertext
		goto err;
```

"Delimiter" after the last field does not matter for good ciphertexts. But with inserted additional fields, it is important. It is possible to generate pieces with all pairs of delimiters, but it increases fuzzing time.

Also if we insert fields, then `valid` checking number of fields would reject such ciphertexts. So the other problem of this approach is the check of number of fields. `pem` does not have such check though. (I would add such check to speed up fuzzing. The speed up would be very small though.)

And such approach would work only if len+data are 2 last fields or there are more such pairs on the end. It would not work for `...$len$aa$bb` where `bb` has fixed length.

With all these drawbacks, this approach found 2 crashes already. I used a dictionary generated this way:
```
$ perl -e 'for $i (1, 10, 100, 1000, 10000) { for $d1 (qw/, . : # $ * @/) { for $d2 (qw/, . : # $ * @/) { printf "%d%s%s%s\n", $i, $d1, "89" x $i, $d2 } } }' > test_fuzz.dic
```

**Comment by AlekseyCherepanov** [comment]
> With all these drawbacks, this approach found 2 crashes already.

4 crashes total (without opencl, ztex and a few formats with slow fuzzing or high memory consumption). But only 1 crash is related to length+data pattern. 2 crashes would happen with insertion of 200 plain digits too. And 1 requires 200 digits plus `$`s around.

I'll file a separate issue for the crashes.

**Comment by AlekseyCherepanov** [comment]
> I'll file a separate issue for the crashes.

#5157

**Comment by AlekseyCherepanov** [comment]
I tried to implement such approach: find number, increase it as possible length, add encoded blob after next char. I found 4 new crashes (added into #5157). Also it can find #5090, while previous approach could not.

But only 2 of new crashes are related to length+data. 2 other crashes would happen with insertion of `0`s repeated different times. Particularly I implemented threating length as hex too. It gave crash in `wowsrp`, but it is not related to length+data.

Good side of this approach is that it is quite fast. All cpu formats complete fuzzing in ~30 seconds.

**Comment by AlekseyCherepanov** [comment]
> It gave crash in `wowsrp`

This one is interesting: I used `strtoll` to get number, it cut long field in hex decoding it as number, then I enlarged next field, and such combo caused crash.

**Comment by AlekseyCherepanov** [comment]
Because I saw a lot of additional crashes caused by plain insertion with unusual size instead of actual length+data mutation, I tried to implement insertion of different lengths. It gave 7 new crashes.

I tried all lengths up to 300 and then up to 20k with some stepping. `ike` is an illustration that stepping is not universal: `ike` crashes with field of length 4096, slight variations do not crash, so it was a coincidence that my stepping added 3840 bytes and together with existing data it was just the right amount.

`rsvp` crashed with 16440 zeros inserted. So big blobs are viable. I have impression that fuzzing of `krb5tgs` becomes slower with longer ciphertexts.

Trying all lengths would be slow. I guess AFL / libFuzzer can trade cycles for reach at better rates. So I am going to prepare PR with my current improvements.

Overall high number of recent crashes and their low complexity suggest that a release requires fuzzing with libFuzzer. (libFuzzer is built-in already.)
